Source/ARTPush.h

```diff
@@ -66,10 +67,22 @@ NS_ASSUME_NONNULL_BEGIN
 + (void)didFailToRegisterForRemoteNotificationsWithError:(NSError *)error realtime:(ARTRealtime *)realtime;
 
 /**
- Activating a device for push notifications and registering it with Ably. The registration process can be performed entirely from the device or from your own server using the optional `ablyPushCustomRegister:deviceDetails:callback` method.
+ Activating a device for push notifications (with requesting user permissions and registering with the APN service).
+ The registration process can be performed on the device or from your own server using the optional ``ablyPushCustomRegister:deviceDetails:callback`` method.
+ */
+- (void)activate:(UNAuthorizationOptions)options;
```

Comment:
I know that this PR isn't ready for review yet, but I happened to see this and wanted to say that I don't think that requesting permissions to display notifications to the user should be part of what the Ably SDK does - that seems like a concern for the application which integrates the SDK, since it might need to be integrated into some flow within the app. The Ably SDK doesn't need these permissions to be able to function - it just needs an APNS device token, which it doesn't need any user permission to get.

Reply:
> since it might need to be integrated into some flow within the app

From docs:
> If you do not request and receive authorization for your app's interactions, the system delivers all remote notifications to your app silently.

I agree too, this creates unwanted scenario of unexpected user interaction. Then I can leave optionless `activate` method just to call `registerForRemoteNotifications:` and the one with the options requesting permissions (as it does now). Or to rename `activate:` to something like `requestAuthorizationAndActivate:` to avoid confusion. I still insist on introducing method for requesting permissions as it greatly reduces initial setup code and thus increases satisfaction of trying the new framework. WDYT @lawrence-forooghian @lukasz-szyszkowski @QuintinWillison

Reply:
I think that we should first fix https://github.com/ably/ably-cocoa/issues/1271, which only requires us to call `-registerForRemoteNotifications`, as @lukasz-szyszkowski said. The user permissions stuff is a separate thing that I think is outside the scope of that issue. I'm personally not in favour of adding further methods to handle the user permissions stuff, but it's something we could discuss as a team separately, I think.

Reply:
Ok, I've created a separate issue for this https://github.com/ably/ably-cocoa/issues/1303

Reply:
I think the concern with this change for me is that it looks like it involves adding a new framework dependency to this SDK, that is [User Notifications](https://developer.apple.com/documentation/usernotifications?language=objc). It begs a few questions for me:

- Do we already require any other App UI oriented frameworks for this SDK?
- Have we ever defined scope/boundaries for this SDK, in respect of baseline/foundational requirements to build and run an app developed against it?

We should be clear as to what the SDK does and doesn't do. It's not a problem to have to spend time communicating to app developers how to integrate their app with the SDK - for example, see [our Flutter PN Documentation](https://github.com/ably/ably-flutter/blob/main/PushNotifications.md) and [our .NET PN Documentation](https://github.com/ably/ably-dotnet/blob/main/PushNotifications.md). 🤯 

Sometimes providing careful guidance that leaves app developers with more flexibility is better, rather than forcing architectural requirements on them. Plus, of course, we don't know how Apple might evolve the `UN` framework going forwards, so we could just be baking in maintenance burden for ourselves going forwards.

Reply:
> Do we already require any other App UI oriented frameworks for this SDK?

Yes, `UIKit` was added by Ricardo back in 2019 to "ARTPush.m":

```
#if TARGET_OS_IOS
#import <UIKit/UIKit.h>
#import "ARTPushActivationStateMachine+Private.h"
#endif
```

> Have we ever defined scope/boundaries for this SDK, in respect of baseline/foundational requirements to build and run an app developed against it? ... Sometimes providing careful guidance that leaves app developers with more flexibility is better, rather than forcing architectural requirements on them.

> Plus, of course, we don't know how Apple might evolve the UN framework going forwards, so we could just be baking in maintenance burden for ourselves going forwards.

I agree, that incorporating ever changing APNs API into ably cocoa is asking for a trouble. On the other hand, many frameworks for providing push notifications do this to make their users life easier. But ably cocoa is more than just push notifications service/framework, so I'm not sure whether we should go into that direction or not. I've created an issue for further discussion of this problem - #1303.

Meanwhile for this PR I'll just make minimum changes required regarding calling `registerForRemoteNotifications`.

Reply:
https://github.com/ably/ably-cocoa/pull/1293/commits/a47ef44df3e93fa2d3208d4a1e342f9f8f834274

Reply:
I'm going to resolve this particular conversation, on behalf of @lawrence-forooghian, as the question it raised has been resolved by @maratal ... however, I have some other questions which I'll raise in a new comment shortly.